File: src/main/java/io/colyseus/Connection.java
package io.colyseus;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.java_websocket.client.WebSocketClient;
import org.java_websocket.drafts.Draft_6455;
import org.java_websocket.handshake.ServerHandshake;
import org.msgpack.jackson.dataformat.MessagePackFactory;

import java.net.URI;
import java.nio.ByteBuffer;
import java.util.LinkedList;
import java.util.Map;

public class Connection extends WebSocketClient {
    private static final int CONNECT_TIMEOUT = 10000;

    interface Listener {
        void onError(Exception e);

        void onClose(int code, String reason, boolean remote);

        void onOpen();

        void onMessage(ByteBuffer bytes);
    }

    private LinkedList<Object[]> _enqueuedCalls = new LinkedList<>();
    private Listener listener;
    private ObjectMapper msgpackMapper;

    Connection(URI uri, Map<String, String> httpHeaders, Listener listener) {
        super(uri, new Draft_6455(), httpHeaders, CONNECT_TIMEOUT);
        System.out.println("Connection()");
        System.out.println("wsUrl is " + uri);
        this.listener = listener;
        this.msgpackMapper = new ObjectMapper(new MessagePackFactory());
        connect();
    }

    void send(Object... data) {
        if (isOpen()) {
            try {
                send(msgpackMapper.writeValueAsBytes(data));
            } catch (Exception e) {
                onError(e);
            }
        } else {
            // WebSocket not connected.
            // Enqueue data to be sent when readyState == OPEN
            this._enqueuedCalls.push(data);
        }
    }

    @Override
    public void onOpen(ServerHandshake handshakedata) {
        if (Connection.this._enqueuedCalls.size() > 0) {
            for (Object[] objects : Connection.this._enqueuedCalls) {
                Connection.this.send(objects);
            }

            // clear enqueued calls.
            Connection.this._enqueuedCalls.clear();
        }
        if (Connection.this.listener != null) Connection.this.listener.onOpen();
    }

    @Override
    public void onMessage(String message) {
//        System.out.println("Connection.onMessage(String message)");
//        System.out.println(message);
    }

    @Override
    public void onClose(int code, String reason, boolean remote) {
        if (Connection.this.listener != null) Connection.this.listener.onClose(code, reason, remote);
    }

    @Override
    public void onError(Exception ex) {
        if (Connection.this.listener != null) Connection.this.listener.onError(ex);
    }

    @Override
    public void onMessage(ByteBuffer buf) {
//        System.out.println("Connection.onMessage(ByteBuffer bytes)");
        if (Connection.this.listener != null) {
//            byte[] bytes = new byte[buf.capacity()];
//            buf.get(bytes, 0, bytes.length);
            Connection.this.listener.onMessage(buf);
        }
    }
}

File: src/main/java/io/colyseus/Connection.kt
package io.colyseus

import com.fasterxml.jackson.databind.ObjectMapper
import org.java_websocket.client.WebSocketClient
import org.java_websocket.drafts.Draft_6455
import org.java_websocket.handshake.ServerHandshake
import org.msgpack.jackson.dataformat.MessagePackFactory
import java.net.URI
import java.nio.ByteBuffer
import java.util.*

private const val CONNECT_TIMEOUT = 10000

class Connection internal constructor(
        uri: URI,
        httpHeaders: Map<String, String>? = null,
) : WebSocketClient(uri, Draft_6455(), httpHeaders, CONNECT_TIMEOUT) {

    var onError: ((e: Exception) -> Unit)? = null
    var onClose: ((code: Int, reason: String?, remote: Boolean) -> Unit)? = null
    var onOpen: (() -> Unit)? = null
    var onMessage: ((bytes: ByteBuffer) -> Unit)? = null

    private val _enqueuedCalls = LinkedList<Any>()
    private val msgpackMapper: ObjectMapper = ObjectMapper(MessagePackFactory())

    fun _send(data: Any) {
        if (isOpen) {
            try {
                val d = msgpackMapper.writeValueAsBytes(data)
//                println("sending... " + Arrays.toString(d))
                send(d)
            } catch (e: Exception) {
                onError(e)
            }
        } else {
            // WebSocket not connected.
            // Enqueue data to be sent when readyState == OPEN
            _enqueuedCalls.push(data)
        }
    }

    override fun onOpen(handshakedata: ServerHandshake) {
        if (!_enqueuedCalls.isEmpty()) {
            _enqueuedCalls.forEach(this::_send)
            _enqueuedCalls.clear()
        }
        onOpen?.invoke()
    }

    override fun onMessage(message: String) {
    }

    override fun onClose(code: Int, reason: String, remote: Boolean) {
        onClose?.invoke(code, reason, remote)
    }

    override fun onError(ex: Exception) {
        onError?.invoke(ex)
    }

    override fun onMessage(buf: ByteBuffer) {
//        println("received: $buf")
        onMessage?.invoke(buf)
    }

    override fun send(data: ByteArray?) {
//        println("sending... " + Arrays.toString(data))
        try {
            super.send(data)
        } catch (e: Exception) {
            onError?.invoke(e)
        }
    }
}